File: cpp-std/demo23a01-thread-local.cpp
/*
THREAD-LOCAL STORAGE
Introduction

The basic way to use thread-local storage
*/


#include <iostream>
#include <string>
#include <thread>
using namespace std;



thread_local string value = "NOT SET";



void routine() {
    cout << value << endl;
}



int main() {
    // Main thread sets value = "APPLE"
    value = "APPLE";
    cout << value << endl;

    // Child thread gets value
    // Expected output: "NOT SET"
    auto th = std::thread(routine);
    th.join();

    return 0;
}

File: csharp/demo/demo23a01-thread-local.cs
/*
 * THREAD-LOCAL STORAGE
 * Introduction
 */
using System;
using System.Threading;



class Demo23A01 : IRunnable
{
    public void run()
    {
        // Main thread sets value = "APPLE"
        MyTask.set("APPLE");
        Console.WriteLine(MyTask.get());

        // Child thread gets value
        // Expected output: "NOT SET"
        new Thread(() =>
        {
            Console.WriteLine(MyTask.get());
        }).Start();
    }


    class MyTask
    {
        private static ThreadLocal<string> data = new ThreadLocal<string>();

        public static string get()
        {
            if (data.Value is null)
            {
                data.Value = "NOT SET";
            }

            return data.Value;
        }

        public static void set(string value)
        {
            data.Value = value;
        }
    }
}
